compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp

```diff
@@ -263,88 +264,166 @@ struct QuantizeDequantizeWeights final : public luci::CircleNodeMutableVisitor<b
   loco::DataType output_type;
   QuantizationGranularity granularity;
 
-  // Quantize and dequantize input tensors of each node
-  bool visit(luci::CircleNode *node)
+private:
+  // Fake quantize weights (Only u8 quantization is supported)
+  void fake_quantize_lwq(luci::CircleConst *weights) const
   {
-    assert(output_type == loco::DataType::U8 || output_type == loco::DataType::S16);
-    LOGGER(l);
-    INFO(l) << "QuantizeDequantizeWeights visit node: " << node->name() << std::endl;
-    auto arity = node->arity();
-    for (uint32_t i = 0; i < arity; i++)
+    // Find min/max per layer
+    float min = std::numeric_limits<float>::max();
+    float max = std::numeric_limits<float>::lowest();
+    for (uint32_t i = 0; i < weights->size<loco::DataType::FLOAT32>(); i++)
     {
-      auto input_node = node->arg(i);
-      auto circle_node = loco::must_cast<luci::CircleNode *>(input_node);
+      auto data = weights->at<loco::DataType::FLOAT32>(i);
+      min = data < min ? data : min;
+      max = data > max ? data : max;
+    }
+    float scaling_factor{0};
+    int64_t zp{0};
+    float nudged_min{0};
+    float nudged_max{0};
+
+    asymmetric_wquant_with_minmax_per_layer(weights, min, max, scaling_factor, zp, nudged_min,
+                                            nudged_max);
+    asymmetric_wdequant_with_minmax_per_layer(weights, scaling_factor, nudged_min);
+    auto quantparam = std::make_unique<CircleQuantParam>();
+    quantparam->min.push_back(nudged_min);
+    quantparam->max.push_back(nudged_max);
+    quantparam->scale.push_back(scaling_factor);
+    quantparam->zerop.push_back(zp);
+    weights->quantparam(std::move(quantparam));
+  }
 
-      // Check if this is already quantized
-      if (is_quantized(circle_node))
-        continue;
+private:
+  // Fake quantize weights (u8/s16 quantization are supported)
+  void fake_quantize_cwq(luci::CircleConst *weights) const
+  {
+    // Find min/max per channel
+    std::vector<float> min;
+    std::vector<float> max;
```

Comment:
Unlike line 271, this is not initialized with `std::numeric_limits<float>::max()` and `std::numeric_limits<float>::lowest()`.

```cpp
    float min = std::numeric_limits<float>::max();
    float max = std::numeric_limits<float>::lowest();
```

And when I looked into `cal_minmax_per_channel` function, I couldn't find the initialization with `max()` and `lowest()` too :(
Is there no need to initialize with those values or did I missed something?

Reply:
This PR just extracts the logic from the existing source code, but the code seems dangerous.

I guess that 0 is set when min/max (std::vector) is resized. We can get the same result (quantized values) with the implicit initialization, but it would be better to specify the initial value (max(), lowest()) when we resize min/max vectors. Nice catch!